src/lib/data/planEntities.ts

```diff
@@ -0,0 +1,236 @@
+import { EntityPrototypeId } from '../../db/models/entityPrototype';
+import { GoverningEntityId } from '../../db/models/governingEntity';
+import { PlanId } from '../../db/models/plan';
+import { PlanEntityId } from '../../db/models/planEntity';
+import { Database } from '../../db/type';
+import { InstanceDataOfModel } from '../../db/util/raw-model';
+import { organizeObjectsByUniqueProperty } from '../../util';
+import { MapOfGoverningEntities } from './governingEntities';
+
+export type ValidatedPlanEntity = {
+  id: PlanEntityId;
+  type: string;
+  customRef: string;
+  description: null | string;
+  /**
+   * List of any entities that this one "supports"
+   * (e.g. specific objectives support strategic objectives)
+   */
+  supports: PlanEntityId[];
+  governingEntity: null | GoverningEntityId;
+};
+
+export type ValidatedPlanEntities = Map<PlanEntityId, ValidatedPlanEntity>;
+
+/**
+ * Get all of the entities for a particular plan,
+ * validate the consistency of data, and return it in a useful format
+ */
+export const getAndValidateAllPlanEntities = async ({
+  database,
+  planId,
+  governingEntities,
+  prototypes,
+}: {
+  database: Database;
+  planId: PlanId;
+  /**
+   * A map from governing entities that include the version to be used when
+   * generating references
+   */
+  governingEntities: MapOfGoverningEntities;
+  version: 'latest';
+  /**
+   * A map of prototypes that **must** include the prototype for governing
+   * entities of this plan.
+   *
+   * Usually the full list of prototypes for a plan is passed in.
+   */
+  prototypes: Map<
+    EntityPrototypeId,
+    InstanceDataOfModel<Database['entityPrototype']>
+  >;
+}): Promise<ValidatedPlanEntities> => {
+  const planEntities = await database.planEntity.find({
+    where: {
+      planId,
+    },
+  });
+  const planEntityIDs = new Set(planEntities.map((pe) => pe.id));
+
+  const pevs = await database.planEntityVersion.find({
+    where: (builder) =>
+      builder
+        .whereIn('planEntityId', [...new Set(planEntities.map((pe) => pe.id))])
```

Comment:
No need to map more than once, we already have this set
```suggestion
        .whereIn('planEntityId', [...planEntityIDs])
```